diff --git a/adbfs.cpp b/adbfs.cpp
index 32e469c..edadc93 100644
--- a/adbfs.cpp
+++ b/adbfs.cpp
@@ -107,6 +107,10 @@ queue<string> adb_shell(const string&, bool);
 queue<string> shell(const string&);
 
 static const char PERMISSION_ERR_MSG[] = ": Permission denied";
+static const char TOUCH_TOYBOX_GNU_DATE_ERR_MSG[] = "touch: bad '@";
+static const char TOUCH_BUSYBOX_GNU_DATE_ERR_MSG[] = "touch: invalid date '@";
+
+
 
 string tempDirPath;
 map<string,fileCache> fileData;
@@ -119,6 +123,7 @@ void invalidateCache(const string& path) {
 
 map<int,bool> filePendingWrite;
 map<string,bool> fileTruncated;
+bool touch_gnu_mode = true;
 
 /**
    Custom options
@@ -782,12 +787,67 @@ static int adb_utimens(const char *path, const struct timespec ts[2]) {
 
     shell_escape_path(path_string);
 
-    queue<string> output;
-    string command = "touch '";
-    command.append(path_string);
-    command.append("'");
+    bool set_atime = true;
+    bool set_mtime = true;
+    struct timespec atime = ts[0];
+    struct timespec mtime = ts[1];
+
+    // Handle UTIME_NOW / UTIME_OMIT
+    struct timespec now;
+    clock_gettime(CLOCK_REALTIME, &now);
+    if (ts[0].tv_nsec == UTIME_NOW) {
+        atime = now;
+    } else if (ts[0].tv_nsec == UTIME_OMIT) {
+        set_atime = false;
+    }
+    if (ts[1].tv_nsec == UTIME_NOW) {
+        mtime = now;
+    } else if (ts[1].tv_nsec == UTIME_OMIT) {
+        set_mtime = false;
+    }
+
+    if (!set_atime && !set_mtime)
+        return 0;
+    string command = "";
+    if (set_atime) {
+        command.append("touch -a -d ");
+        command.append(format_touch_time(atime, touch_gnu_mode));
+        command.append(" '");
+        command.append(path_string);
+        command.append("'");
+        if (set_mtime)
+            command.append(" && ");
+    }
+    if (set_mtime) {
+        command.append("touch -m -d ");
+        command.append(format_touch_time(mtime, touch_gnu_mode));
+        command.append(" '");
+        command.append(path_string);
+        command.append("'");
+    }
     cout << command<<"\n";
-    adb_shell(command);
+    queue<string> output;
+    output = adb_shell(command, true);
+    if (!output.empty()) {
+        string front = output.front();
+        while (!output.empty()) {
+            cout << output.front() << "\n";
+            output.pop();
+        }
+        bool date_failed = front.length() > sizeof(TOUCH_TOYBOX_GNU_DATE_ERR_MSG)
+                && !front.compare(0, sizeof(TOUCH_TOYBOX_GNU_DATE_ERR_MSG) - 1, TOUCH_TOYBOX_GNU_DATE_ERR_MSG);
+        date_failed = date_failed || (front.length() > sizeof(TOUCH_BUSYBOX_GNU_DATE_ERR_MSG)
+                && !front.compare(0, sizeof(TOUCH_BUSYBOX_GNU_DATE_ERR_MSG) - 1, TOUCH_BUSYBOX_GNU_DATE_ERR_MSG));
+
+        if (date_failed) {
+            if (!touch_gnu_mode)
+                return -ENOSYS;
+
+            cout << "Touch doesn't seems to support GNU dates with nanoseconds support. Switching to legacy mode.\n";
+            touch_gnu_mode = false;
+            return adb_utimens(path, ts);
+        }
+    }
 
     // If we forgot to mount -o rescan then we can remount and touch to trigger the scan.
     if (adbfs_conf.rescan) adb_rescan_file(path_string);
@@ -804,8 +864,9 @@ static int adb_truncate(const char *path, off_t size) {
     string_replacer(path_string,"/","-");
     local_path_string.append(path_string);
     path_string.assign(path);
+    string local_path_string_escaped = local_path_string;
     shell_escape_path(path_string);
-    shell_escape_path(local_path_string);
+    shell_escape_path(local_path_string_escaped);
 
 
     queue<string> output;
@@ -817,7 +878,7 @@ static int adb_truncate(const char *path, off_t size) {
     output = adb_shell(command);
     vector<string> output_chunk = make_array(output.front());
     if (output_chunk[0][0] == '/'){
-        adb_pull(path_string,local_path_string);
+        adb_pull(path_string,local_path_string_escaped);
     }
 
     fileTruncated[path_string] = true;
diff --git a/utils.h b/utils.h
index 223785b..58b2a45 100644
--- a/utils.h
+++ b/utils.h
@@ -48,6 +48,8 @@
 #include <vector>
 #include <map>
 #include <unistd.h>
+#include <ctime>
+#include <iomanip>
 
 using namespace std;
 
@@ -177,3 +179,31 @@ queue<string> exec_command(const string& command)
     return output;
 }
 
+/**
+   Convert timespec to touch parameter format
+
+   gnu_format allows to specify up to nanoseconds.
+   https://www.gnu.org/software/coreutils/manual/html_node/General-date-syntax.html
+   Toybox's touch used on Android supports the GNU extension for the nanoseconds since 0.8.1 : "@%s.%N"
+   Toybox was begin used instead of Busybox in Android 6.
+
+   @param t the time to be converted
+   @param gnu_format if touch supports GNU time format
+ */
+static string format_touch_time(const struct timespec& t, bool gnu_format) {
+    ostringstream ss;
+    if (gnu_format) {
+        ss << "@"
+            << t.tv_sec
+            << "."
+            << setw(9) << setfill('0') << t.tv_nsec;
+        return ss.str();
+    } else {
+        // It could be possible to convert here the timezone instead of creating a more complex shell comamand.
+        ss << "`date -ud "
+            << "@" << t.tv_sec  << "."
+            << setw(9) << setfill('0') << t.tv_nsec
+            << " +%Y-%m-%dT%H:%M:%S`";
+    }
+    return ss.str();
+}
